compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -1552,12 +1576,72 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 }
 
                 err.emit_unless(unsized_return);
+                if ret_vec.len() > 0 {
+                    if let Some(expr) = expression {
+                        let mut addtion_err = fcx.tcx.sess.diagnostic().span_note_diag(expr.span, "the function expects a value to always be returned, but loops might run zero times");
+                        for i in 0..ret_vec.len() {
+                            //let ret_expr = fcx.tcx.hir().get(ret_vec[i]);
+                            addtion_err.span_label(
+                                fcx.tcx.hir().span(ret_vec[i]),
+                                "if the loop doesn't execute, this value would never get returned"
+                                    .to_string(),
+                            );
+                        }
+                        let loop_span = if let hir::ExprKind::Loop(_, _, _, loop_span) = expr.kind {
+                            loop_span
+                        } else {
+                            todo!()
+                        };
+                        addtion_err.span_label(
+                            loop_span,
+                            "this might have zero elements to iterate on".to_string(),
+                        );
+                        addtion_err.help("return a value for the case when the loop has zero elements to iterate on, and consider changing the return type to account for that possibility");
```

Comment:
I think the original issue #98982 was asking to make this into a suggestion, i.e. have the compiler suggest wrapping the return values in `Some(..)` and then returning `None`

Reply:
I think we need better machinery for default values (damn! I just realized that we could opportunistically programatically get that from the `Default` impl of types!), I'm ok with a `help` for now.